File: src/nuitrack_bridge/types/depth_frame.cc
#include "nuitrack_bridge/types/depth_frame.h"
#include "nuitrack-rs/src/nuitrack_bridge/types/depth_frame.rs.h"
#include "nuitrack/types/DepthFrame.h"
#include "nuitrack/types/Frame.h"

namespace nuitrack_bridge::depth_frame {

    // --- DepthFrame functions ---
    int32_t getRows(const DepthFrame& frame) {
        return static_cast<int32_t>(frame.getRows());
    }

    int32_t getCols(const DepthFrame& frame) {
        return static_cast<int32_t>(frame.getCols());
    }

    uint64_t getID(const DepthFrame& frame) {
        return frame.getID();
    }

    uint64_t getTimestamp(const DepthFrame& frame) {
        return frame.getTimestamp();
    }

    rust::Slice<const uint16_t> getData(const DepthFrame& frame) {
        const uint16_t* data_ptr = frame.getData();
        size_t num_pixels = static_cast<size_t>(frame.getRows()) * static_cast<size_t>(frame.getCols());
        return rust::Slice<const uint16_t>{data_ptr, num_pixels};
    }
}

File: src/nuitrack_bridge/types/depth_frame.rs
#[cxx::bridge(namespace = "nuitrack_bridge::depth_frame")]
pub mod ffi {

    #[namespace = "tdv::nuitrack"]
    unsafe extern "C++" {
        type DepthFrame;
    }

    impl SharedPtr<DepthFrame> {}

    unsafe extern "C++" {
        include!("nuitrack_bridge/types/depth_frame.h");

        #[cxx_name = "getRows"]
        pub fn rows(frame: &DepthFrame) -> Result<i32>;

        #[cxx_name = "getCols"]
        pub fn cols(frame: &DepthFrame) -> Result<i32>;

        #[cxx_name = "getID"]
        pub fn id(frame: &DepthFrame) -> Result<u64>;

        #[cxx_name = "getTimestamp"]
        pub fn timestamp(frame: &DepthFrame) -> Result<u64>;

        #[cxx_name = "getData"]
        pub fn data(frame: &DepthFrame) -> Result<&[u16]>;
    }
}

unsafe impl Send for ffi::DepthFrame {}
unsafe impl Sync for ffi::DepthFrame {}






